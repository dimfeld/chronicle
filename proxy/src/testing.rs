use std::{
    sync::{atomic::AtomicUsize, Arc},
    time::Duration,
};

use error_stack::{Report, ResultExt};

use crate::{
    format::{ChatChoice, ChatMessage, ChatResponse, UsageResponse},
    providers::{
        ChatModelProvider, ProviderError, ProviderErrorKind, SendRequestOptions,
        SynchronousProviderResponse,
    },
    Error,
};

#[derive(Debug)]
pub enum TestFailure {
    Transient,
    Timeout,
    BadRequest,
    RateLimit,
    Auth,
    TransformingRequest,
    TransformingResponse,
}

#[derive(Debug)]
pub struct TestProvider {
    pub name: String,
    /// Fail requests
    pub fail: Option<TestFailure>,
    pub response: String,
    pub fail_times: usize,
    pub calls: AtomicUsize,
}

impl Default for TestProvider {
    fn default() -> Self {
        Self {
            name: "test".to_string(),
            fail: None,
            response: "A response".to_string(),
            fail_times: usize::MAX,
            calls: AtomicUsize::new(0),
        }
    }
}

impl Into<Arc<dyn ChatModelProvider>> for TestProvider {
    fn into(self) -> Arc<dyn ChatModelProvider> {
        Arc::new(self)
    }
}

#[async_trait::async_trait]
impl ChatModelProvider for TestProvider {
    fn name(&self) -> &str {
        "test"
    }

    fn label(&self) -> &str {
        "Test"
    }

    async fn send_request(
        &self,
        options: SendRequestOptions,
    ) -> Result<SynchronousProviderResponse, Report<Error>> {
        let current_call = self
            .calls
            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);

        if current_call < self.fail_times {
            match self.fail {
                Some(TestFailure::Transient) => Err(ProviderErrorKind::Server),
                Some(TestFailure::Timeout) => Err(ProviderErrorKind::Timeout),
                Some(TestFailure::BadRequest) => Err(ProviderErrorKind::BadInput),
                Some(TestFailure::RateLimit) => {
                    Err(ProviderErrorKind::RateLimit { retry_after: None })
                }
                Some(TestFailure::Auth) => Err(ProviderErrorKind::AuthRejected),
                Some(TestFailure::TransformingRequest) => return Err(Error::TransformingRequest)?,
                Some(TestFailure::TransformingResponse) => {
                    return Err(Error::TransformingResponse)?
                }
                None => Ok(()),
            }
            .map_err(|kind| ProviderError {
                kind,
                status_code: None,
                body: None,
                latency: Duration::from_millis(500),
            })
            .change_context(Error::ModelError)?;
        }

        Ok(SynchronousProviderResponse {
            model: options.body.model.clone().unwrap_or_default(),
            body: ChatResponse {
                created: 1,
                model: options.body.model.clone(),
                system_fingerprint: None,
                choices: vec![ChatChoice {
                    index: 0,
                    message: ChatMessage {
                        role: Some("assistant".to_string()),
                        content: Some(self.response.clone()),
                        tool_calls: Vec::new(),
                        name: None,
                    },
                    finish_reason: "stop".to_string(),
                }],
                usage: UsageResponse {
                    prompt_tokens: Some(1),
                    completion_tokens: Some(2),
                    total_tokens: Some(3),
                },
            },
            meta: None,
            latency: Duration::from_millis(500),
        })
    }

    fn is_default_for_model(&self, _model: &str) -> bool {
        false
    }
}
