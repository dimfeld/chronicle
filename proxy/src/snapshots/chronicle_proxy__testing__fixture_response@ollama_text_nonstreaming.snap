---
source: proxy/src/testing.rs
expression: response
---
{
  "request_info": {
    "id": "00000000-0000-0000-0000-000000000000",
    "provider": "ollama",
    "model": "me/a-test-model",
    "num_retries": 0,
    "was_rate_limited": false
  },
  "response_info": {
    "meta": {
      "eval_duration": 4889146000,
      "load_duration": 3278875,
      "prompt_eval_duration": 162740000
    },
    "model": "llama3"
  },
  "was_streaming": false,
  "num_chunks": 1,
  "response": {
    "created": 0,
    "model": "llama3",
    "system_fingerprint": null,
    "choices": [
      {
        "index": 0,
        "message": {
          "role": "assistant",
          "content": "A beloved topic!\n\nLassie is a fictional Rough Collie dog character created by Eric Knight in his 1940 novel \"Lassie Come-Home.\" The story follows a loyal and courageous golden-colored collie named Lassie, who loves her human family, particularly Timmy Martin. After being separated from her family during a tragic accident, Lassie embarks on an incredible journey to find them.\n\nThe character has since become a cultural icon through various adaptations:\n\n1. **Novels**: Eric Knight wrote six more \"Lassie\" novels between 1940 and 1957.\n2. **Films**: The first film, \"Lassie Come-Home,\" was released in 1943, followed by over 30 more movies featuring Lassie from the 1940s to the 1990s.\n3. **Television**: A popular TV series aired from 1954 to 1974, with a total of 574 episodes. The show starred Lassie as a loyal and heroic canine companion, often helping her human friends in need.\n4. **Radio**: Lassie had her own radio show, \"The Lassie Story,\" which ran from 1947 to 1953.\n\nThroughout the years, Lassie has become synonymous with loyalty, courage, and friendship, making her one of the most beloved and recognizable canine characters in popular culture."
        },
        "finish_reason": "stop"
      }
    ],
    "usage": {
      "prompt_tokens": null,
      "completion_tokens": 287,
      "total_tokens": null
    }
  }
}
